src/parser.cpp

```diff
@@ -562,11 +562,14 @@ job_t *parser_t::job_get(job_id_t id) {
     return NULL;
 }
 
-job_t *parser_t::job_get_from_pid(int pid) {
+job_t *parser_t::job_get_from_pid(pid_t pid) {
     job_iterator_t jobs;
     job_t *job;
+
+    pid_t pgid = getpgid(pid);
+
     while ((job = jobs.next())) {
-        if (job->pgid == pid) return job;
+        if (job->pgid == pgid) return job;
```

Comment:
This seems correct.

`sudo echo lol | grep lol` should be one job with two processes (plus keepalive), and they should have the same pgid. In that case, whether you ask for the job that contains the "sudo" process or the "grep" process, you get the same one. If you compared "pgid == pid" instead, it would only succeed with one.

AFAIU right now, the "sudo" process would get a different PGID from the "grep" process, but they would still share the same job. In that case, nothing would break with this change, but it still wouldn't fix the case where you ask for the process that doesn't have the same pgid as our job object.

If they do currently get different job objects, then all is fine and dandy except EVERYTHING IS BROKEN AND JOB CONTROL IS HORRIBLE AND MY HOUSE IS ON FIRE.